### PR TITLE
Add kpod exec

### PR DESCRIPTION
**Pull request by baude**
Run a command in a running container

Signed-off-by: baude <bbaude@redhat.com>

### CODE REVIEW

File: cmd/kpod/exec.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+	PID in the container is running.
+`
+
+	execCommand = cli.Command{
+		Name:        "exec",
+		Usage:       "Run a command in a running container",
+		Description: execDescription,
+		Flags:       execFlags,
+		Action:      execCmd,
+		ArgsUsage:   "[OPTIONS} CONTAINER-NAME COMMAND",
```
Comment (rhatdan) [inline cmd/kpod/exec.go]: This line is wrong.

File: cmd/kpod/exec.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	}
+)
+
+func execCmd(c *cli.Context) error {
+	args := c.Args()
+	detachFlag := c.Bool("detach")
```
Comment (rhatdan) [inline cmd/kpod/exec.go]: No need for variable, it is only used once.

File: cmd/kpod/exec.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	envFlags := c.StringSlice("env")
```
Comment (rhatdan) [inline cmd/kpod/exec.go]: No need for variable, it is only used once.

File: cmd/kpod/exec.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+	ttyFlag := c.Bool("tty")
```
Comment (rhatdan) [inline cmd/kpod/exec.go]: No need for variable, it is only used once.

File: cmd/kpod/exec.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+	userFlag := c.String("user")
```
Comment (rhatdan) [inline cmd/kpod/exec.go]: No need for variable, it is only used once.

File: libkpod/exec.go [hunk 6]
```diff
@@ -0,0 +1,29 @@
+package libkpod
```
Comment (rhatdan) [inline libkpod/exec.go]: No more libkpod

**Comment by rhatdan** [comment]
What about 
--detach-keys=
--privileged=true|false

**Comment by rhatdan** [comment]
Need changes in README.md and docs/kpod.1.md

File: libkpod/exec.go (hunk 6 above)
Comment (baude) [reply libkpod/exec.go]: @rhatdan we have to use libkpod until libpod is ready for container related stuff, which it isn't.  I talked to @mheon about this and we considered progress on the commands to be more important ATM.

**Comment by baude** [comment]
@rhatdan runc does not honor either of those switches ... and also --interactive.

File: cmd/kpod/exec.go (hunk 2 above)
Comment (baude) [reply cmd/kpod/exec.go]: fixed.

File: cmd/kpod/exec.go (hunk 3 above)
Comment (baude) [reply cmd/kpod/exec.go]: fixed

File: cmd/kpod/exec.go (hunk 4 above)
Comment (baude) [reply cmd/kpod/exec.go]: fixed

File: cmd/kpod/exec.go (hunk 5 above)
Comment (baude) [reply cmd/kpod/exec.go]: fixed

**Comment by rhatdan** [comment]
@baude Then what does docker do?

File: libkpod/exec.go (hunk 6 above)
Comment (mheon) [reply libkpod/exec.go]: Especially considering how minimal the backend code on this is. We need to support additional functionality (like attaching to exec'd processes) at some later point, but this gets us some form of exec with almost no code on the backend.

File: cmd/kpod/exec.go (hunk 5 above)
Comment (rhatdan) [reply cmd/kpod/exec.go]: Well we can merge this in, but we need to accellerate the move to libpod.

File: oci/oci.go [hunk 7]
```diff
@@ -716,3 +716,30 @@ func (r *Runtime) UnpauseContainer(c *Container) error {
...
+		commandArgs = append(commandArgs, "-u", user)
+	}
+	commandArgs = append(commandArgs, c.id)
+	commandArgs = append(commandArgs, command...)
+	logrus.Debug("Running command: %s", strings.Join(commandArgs, " "))
+	c.opLock.Lock()
+	defer c.opLock.Unlock()
+	err := utils.ExecCmdWithStdStreams(os.Stdin, os.Stdout, os.Stderr, r.Path(c), commandArgs...)
+	return err
+	return nil
```
Comment (TomSweeneyRedHat) [inline oci/oci.go]: I think you've an extra return here.

File: cmd/kpod/exec.go [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+	"github.com/kubernetes-incubator/cri-o/libkpod"
+	"github.com/pkg/errors"
+	"github.com/urfave/cli"
+)
+
+var (
+	execFlags = []cli.Flag{
+		cli.BoolFlag{
+			Name:  "detach, d",
+			Usage: "detached mode: command will run in the background",
```
Comment (TomSweeneyRedHat) [inline cmd/kpod/exec.go]: The usage statements some are capitalized to start, some are not.  We should be consistent.   However I just noted that we upper case in some commands and lower case in other kpod command.  @rhatdan these should be all lower to start right?  i.e. "allocate a pseudo TTY".

Comment (rhatdan) [reply cmd/kpod/exec.go]: SGTM

File: libkpod/exec.go (hunk 6 above)
Comment (rhatdan) [reply libkpod/exec.go]: Ok I give in, but we really need to start moving to libpod.

**Comment by rhatdan** [comment]
runc exec
has 

```
       --console value              specify the pty slave path for use with the container
          --cwd value                  current working directory in the container
          --env value, -e value        set environment variables
          --tty, -t                    allocate a pseudo-TTY
          --user value, -u value       UID (format: <uid>[:<gid>])
          --process value, -p value    path to the process.json
          --detach, -d                 detach from the container's process
          --pid-file value             specify the file to write the process id to
          --process-label value        set the asm process label for the process commonly used with selinux
          --apparmor value             set the apparmor profile for the process
          --no-new-privs               set the no new privileges value for the process
          --cap value, -c value        add a capability to the bounding set for the process
          --no-subreaper               disable the use of the subreaper used to reap reparented processes
```
--privileged would be to specify nothing for process-label, apparmor, --no-new-privs and to add all capabilities.


When you are just running without --privilege, you need to inspect the container and set those fields correctly.

--process-label=SELINUXLABEL - If Specified
--APPARMOR=APPARMORPROFILE - if specified

**Comment by rhatdan** [comment]
@baude You should check to make sure that the label of the process you are injecting into the container matches the labels of the other processes in the container.  Similarly use /usr/bin/pscap to look at the list of capabilties associated with the container processes and the new process.
Finally we need to make sure the cgroup matches, UID Matches etc.

**Comment by rhatdan** [comment]
@baude Lets move this over to libpod.

**Comment by runcom** [comment]
I'm closing this because of the kpod move, lemme know if I have to reopen

**Pull request closed without merging**